doc/workflows/anaconda-webui.rst

```diff
@@ -0,0 +1,225 @@
+
+    execution:
+        type: anaconda-webui
+        automation_data:
+            script_file: ./ui/webui/test/integration/default.py
+            test_case: DefaultInstallation
+
+Example of execution section with all options set::
+
+    execution:
+        type: anaconda-webui
+        automation_data:
+            script_file: ./check-navigation.py
+            test_case: TestNavigation
+            test_repo: my-special-tests
+            additional_repos: ['AppStream', 'CRB']
+            kernel_cmdline: 'nosmt'
+            webui_startup_timeout: 15
+
+..  _hypervisor:
+
+Hypervisor
+^^^^^^^^^^
+The workflow is using virtual machines, so it needs access to a system with libvirtd running.
+That can be eighter localhost or remote system accessible via ssh. To prepare the system to
+be used by this workflow: 
+
+1. Install `libvirt` and `qemu-kvm` packges
+2. Start `libvirtd` service.
+3. Make sure the system can be accessed by permian without password (set authorized_keys for selected user).
+4. If the user is not root, add it to the group `libvirt`.
+
+
+Test scripts
+^^^^^^^^^^^^
+Anaconda WebUI integration tests are python scripts that use cockpit's test framework
+and wrapper methods that make it easy to interact with the Web UI and run commands
+on the machine during installation and after reboot.
+
+More information can be found in `Anaconda documentation <https://anaconda-installer.readthedocs.io/en/latest/testing.html#anaconda-web-ui-tests>`_.
+
+Execution
+---------
+
+Run a test case from local anaconda repo
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+We have special command that make this the easiest way to run WebUI test.
+
+1. Follow the guide on page :ref:`Quick start<Quick start>` and get Permian
+   running without an container.
+2. By default your system will be used as :ref:`hypervisor`, so make sure libvirtd is running.
+3. Clone anaconda repository::
+
+    git clone -–depth 1 https://github.com/rhinstaller/anaconda.git
+
+4. (optional) create new test case and test script file in the anaconda repository
+   we just cloned.
+
+   1. Create new file `anaconda/ui/webui/test/integration/my_new_test.tc.yaml` with
+      following content::
+
+        name: My new test
+        description: Example test case
+        author: {you e-mail here}
+        tags:
+          - anaconda
+        priority: 1
+        execution:
+          type: anaconda-webui
+          automation_data:
+            script_file: ./ui/webui/test/integration/my_new_test.py
+            test_case: MyNewTest
+
+   2. Copy any existing test script to `anaconda/ui/webui/test/integration/my_new_test.py`,
+      good examples are listed in `Integration tests examples <https://anaconda-installer.readthedocs.io/en/latest/testing.html#integration-tests-examples>`.
+   3. Change test class name in the test script to `MyNewTest`
+
+5. Run the new test case::
+
+    PYTHONPATH=../tplib ./pipeline run_awebui_tc ../anaconda 'My new test'
```

Comment:
I forgot to test this, the tc.yaml is missing a part. Same goes for the error bellow.

Reply:
Unfortunately, I hit another issue. The command starts, but when it tries to run `virt-install`, it gets stuck on asking for authentication forever. A popup window keeps popping up asking for password, but only an error `error: failed to get domain 'anaconda-webui--...` is printed. See the output below:
```
$ PYTHONPATH=../tplib ./pipeline run_awebui_tc ../anaconda 'My new test'
 * Serving Flask app 'libpermian.webui.server'
 * Debug mode: off
INFO:libpermian.webui.callbacks(Thread-2 (signalWhenStarted)):WebUI started at: http://192.168.1.102:45897/
Cloning into '/tmp/tmpih11xaho'...
remote: Enumerating objects: 1469, done.
remote: Counting objects: 100% (1469/1469), done.
remote: Compressing objects: 100% (1289/1289), done.
remote: Total 1469 (delta 264), reused 639 (delta 96), pack-reused 0
Receiving objects: 100% (1469/1469), 7.22 MiB | 2.51 MiB/s, done.
Resolving deltas: 100% (264/264), done.
Cloning into '/var/tmp/pipeline_awebui_y0pe_k6q/anaconda/ui/webui/bots'...
remote: Enumerating objects: 243, done.
remote: Counting objects: 100% (243/243), done.
remote: Compressing objects: 100% (212/212), done.
remote: Total 243 (delta 24), reused 139 (delta 9), pack-reused 0
Receiving objects: 100% (243/243), 218.36 KiB | 113.00 KiB/s, done.
Resolving deltas: 100% (24/24), done.
INFO:libpermian.plugins.anaconda_webui(Thread-5):Running: ['virt-install', '--connect', 'qemu:///system', '--autoconsole', 'text', '-n', 'anaconda-webui--5907613892168106095', '--os-variant', 'rhel-unknown', '--location', 'https://fedorapeople.org/groups/anaconda/webui_permian_tests/sources/periodic/x86_64/,kernel=images/pxeboot/vmlinuz,initrd=images/pxeboot/initrd.img', '--memory', '4096', '--vcpus', '2', '--disk', 'size=10', '--serial', 'pty', '--extra-args', 'inst.sshd inst.webui inst.webui.remote inst.graphical console=ttyS0 inst.stage2=https://fedorapeople.org/groups/anaconda/webui_permian_tests/sources/periodic/x86_64/ inst.geoloc=0']
error: failed to get domain 'anaconda-webui--5907613892168106095'
error: failed to get domain 'anaconda-webui--5907613892168106095'
```
Sometimes, it is even not possible to interrupt the process via ctrl-C.

When I tried to run the `virt-install` command separately, it asked for authentication just once and then continued as expected.

Reply:
Yeah, it would be worth documenting that one needs to have proper privileges. Unfortunately, this is hard to document as this depends if local or remote hypervisor is used.
In your case (to unblock the review), if you're in libvirt group, it should start working without the auth prompt.

Reply:
There should be no authentication. That sounds like you are not in the group libvirt. It's mentioned on the page but I will repeat it in the execution section.

Reply:
It would be great to point the user to all the required setup (e.g. the 'Hypervisor' section), because the examples are often the first thing people are looking for, skipping all the other details (as I did).

Reply:
I don't know how the libvirt service usually behaves, but there isn't any other setup required. It's normal for the workflow to print error: failed to get domain while it waits for virt-install to download kernel, initrd and start the VM. I could hide that error or replace it with more informative message, but that should be done in different PR.

Reply:
A more informative message would highly appreciated. I did not know this kind of error is normal and because my sample test always fails (I don't know why), I thought it is related (I still suspect is might).

This is my test setup (just followed the guide):

anaconda/ui/webui/test/integration/my_new_test.tc.yaml:
```yaml
name: My new test
description: Short description of the test case
author: pberanek@redhat.com
tags:
  - anaconda
priority: 1
execution:
  type: anaconda-webui
  automation_data:
    script_file: ./ui/webui/test/integration/my_new_test.py
    test_case: MyNewTest
instructions:
  steps:
    - step: Describe your test steps here
```

anaconda/ui/webui/test/integration/my_new_test.py
```python
#!/usr/bin/python3
#
# Copyright (C) 2022 Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify it
# under the terms of the GNU Lesser General Public License as published by
# the Free Software Foundation; either version 2.1 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program; If not, see <http://www.gnu.org/licenses/>.

import os
import sys

# import Cockpit's machinery for test VMs and its browser test API
TEST_DIR = os.environ['WEBUI_TEST_DIR']
sys.path.append(TEST_DIR)
sys.path.append(os.path.join(TEST_DIR, "common"))
sys.path.append(os.path.join(TEST_DIR, "helpers"))
sys.path.append(os.path.join(os.path.dirname(TEST_DIR), "bots/machine"))

from integration import IntegrationTest  # pylint: disable=import-error
from testlib import test_main  # pylint: disable=import-error

class MyNewTest(IntegrationTest):
    def test_default_installation(self):
        self.run_integration_test()

if __name__ == '__main__':
    test_main()
```

Pipeline output:
```
$ PYTHONPATH=../tplib ./pipeline run_awebui_tc ../anaconda 'My new test'
 * Serving Flask app 'libpermian.webui.server'
 * Debug mode: off
INFO:libpermian.webui.callbacks(Thread-2 (signalWhenStarted)):WebUI started at: http://192.168.1.101:59219/
STEP 1/2: FROM quay.io/centos/centos:stream9
... (long output) ...
Successfully tagged localhost/anaconda-webui:latest
18a63e37ac95966463d40376dee699aae803b57fe75dfb40d65f6658857831d2
Cloning into '/tmp/tmptez8usep'...
remote: Enumerating objects: 1469, done.
remote: Counting objects: 100% (1469/1469), done.
remote: Compressing objects: 100% (1307/1307), done.
remote: Total 1469 (delta 263), reused 619 (delta 78), pack-reused 0
Receiving objects: 100% (1469/1469), 7.23 MiB | 3.56 MiB/s, done.
Resolving deltas: 100% (263/263), done.
Cloning into '/var/tmp/pipeline_awebui_4auyx5hc/anaconda/ui/webui/bots'...
remote: Enumerating objects: 240, done.
remote: Counting objects: 100% (240/240), done.
remote: Compressing objects: 100% (210/210), done.
remote: Total 240 (delta 23), reused 134 (delta 8), pack-reused 0
Receiving objects: 100% (240/240), 217.75 KiB | 917.00 KiB/s, done.
Resolving deltas: 100% (23/23), done.
INFO:libpermian.plugins.anaconda_webui(Thread-5):Running: ['virt-install', '--connect', 'qemu:///system', '--autoconsole', 'text', '-n', 'anaconda-webui--3034567895834912952', '--os-variant', 'rhel-unknown', '--location', 'https://fedorapeople.org/groups/anaconda/webui_permian_tests/sources/periodic/x86_64/,kernel=images/pxeboot/vmlinuz,initrd=images/pxeboot/initrd.img', '--memory', '4096', '--vcpus', '2', '--disk', 'size=10', '--serial', 'pty', '--extra-args', 'inst.sshd inst.webui inst.webui.remote inst.graphical console=ttyS0 inst.stage2=https://fedorapeople.org/groups/anaconda/webui_permian_tests/sources/periodic/x86_64/ inst.geoloc=0']
error: failed to get domain 'anaconda-webui--3034567895834912952'
Untagged: localhost/anaconda-webui:latest
Deleted: 18a63e37ac95966463d40376dee699aae803b57fe75dfb40d65f6658857831d2
INFO:libpermian.pipeline(MainThread):All execution and reporting is done. Performing other post-reporting and shutdown activities.
INFO:libpermian.webui.callbacks(MainThread):Generating static WebUI
INFO:libpermian.webui.callbacks(MainThread):Static WebUI generation is complete
```

When I check files created by permian in the working dir (running `python3 -m http.server`), test result is `FAIL`.
Logs > workflow:
```
2023-05-17 14:41:46 - Waiting for setup lock
2023-05-17 14:41:46 - Running setup
2023-05-17 14:43:41 - Clonning common repositories
2023-05-17 14:43:47 - Installing NPM dependencies
2023-05-17 14:43:50 - Waiting for VM slot
2023-05-17 14:43:50 - Changing state to: "started" with result: "None"
2023-05-17 14:43:50 - Starting VM
2023-05-17 14:43:50 - Waiting for IP
2023-05-17 14:44:50 - Waiting for WebUI
2023-05-17 14:47:00 - Running test MyNewTest
2023-05-17 14:47:00 - Changing state to: "running" with result: "None"
2023-05-17 14:47:00 - Running: ./ui/webui/test/integration/my_new_test.py MyNewTest --browser 192.168.122.236:9090 --machine 192.168.122.236:22
2023-05-17 14:47:13 - Test finished
2023-05-17 14:47:13 - Changing state to: "complete" with result: "FAIL"
2023-05-17 14:47:13 - Teardown
2023-05-17 14:47:13 - Adding log file anaconda.log
2023-05-17 14:47:13 - Adding log file packaging.log
2023-05-17 14:47:13 - Adding log file storage.log
2023-05-17 14:47:13 - Adding log file dbus.log
2023-05-17 14:47:13 - Adding log file syslog
```

Logs > virt-install:
```
... (long output) ...
anaconda 39.9-1.fc37 for Rawhide v41 (pre-release) started.
 * installation log files are stored in /tmp during the installation
 * shell is available on TTY2 and in second TMUX pane (ctrl+b, then press 2)
 * when reporting a bug add logs from /tmp as separate text/plain attachments
[?7727hQueued start job for default target default.target.
Startup finished in 264ms.
Waiting for all threads to finish...

Running text console command: virsh --connect qemu:///system console anaconda-webui--3034567895834912952
Domain creation completed.
User stopped the VM. Not rebooting.
```

The test failed almost immediately. What am I doing wrong?

Reply:
Looks like Permian and workflow worked. This is problem in the test or Anaconda, check output.txt.

Reply:
Missing dependencies? (a Node package):
```
2023-05-17 14:47:13 - # ----------------------------------------------------------------------
# test_default_installation (__main__.MyNewTest)
[TEST STEP] open
[0517/144711.093853:WARNING:discardable_shared_memory_manager.cc(197)] Less than 64MB of free space in temporary directory for shared memory files: 62
[0517/144711.093905:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[0517/144711.100964:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory
[0517/144711.100992:ERROR:bus.cc(399)] Failed to connect to the bus: Failed to connect to socket /run/dbus/system_bus_socket: No such file or directory

DevTools listening on ws://127.0.0.1:9385/devtools/browser/a7b8ed1c-6520-46c7-ae58-d42df7a03e80
[0517/144711.105857:WARNING:bluez_dbus_manager.cc(247)] Floss manager not present, cannot set Floss enable/disable.
[0517/144711.161390:WARNING:sandbox_linux.cc(393)] InitializeSandbox() called with multiple threads in process gpu-process.
node:internal/errors:478
    ErrorCaptureStackTrace(err);
    ^

Error [ERR_UNKNOWN_BUILTIN_MODULE]: No such built-in module: node:readline/promises
[90m    at new NodeError (node:internal/errors:387:5)[39m
[90m    at ESMLoader.builtinStrategy (node:internal/modules/esm/translators:259:11)[39m
[90m    at ESMLoader.moduleProvider (node:internal/modules/esm/loader:468:14)[39m {
  code: [32m'ERR_UNKNOWN_BUILTIN_MODULE'[39m
}
Traceback (most recent call last):
  File "/usr/lib64/python3.9/unittest/case.py", line 550, in _callTestMethod
    method()
  File "/root/workdir/anaconda/./ui/webui/test/integration/my_new_test.py", line 33, in test_default_installation
    self.run_integration_test()
  File "/root/workdir/anaconda/ui/webui/test/helpers/integration.py", line 109, in run_integration_test
    self._installer.open()
  File "/root/workdir/anaconda/ui/webui/test/helpers/step_logger.py", line 40, in wrapper
    result = function(*args, **kwargs)
  File "/root/workdir/anaconda/ui/webui/test/helpers/installer.py", line 92, in open
    self.browser.open(f"/cockpit/@localhost/anaconda-webui/index.html#/{step}")
  File "/root/workdir/anaconda/ui/webui/test/common/testlib.py", line 235, in open
    self._set_window_size(size[0], size[1])
  File "/root/workdir/anaconda/ui/webui/test/common/testlib.py", line 966, in _set_window_size
    self.cdp.invoke("Emulation.setDeviceMetricsOverride",
  File "/root/workdir/anaconda/ui/webui/test/common/cdp.py", line 205, in invoke
    res = self.command(cmd)
  File "/root/workdir/anaconda/ui/webui/test/common/cdp.py", line 215, in command
    self.start()
  File "/root/workdir/anaconda/ui/webui/test/common/cdp.py", line 327, in start
    self.invoke("Page.addScriptToEvaluateOnNewDocument", source=src, no_trace=True)
  File "/root/workdir/anaconda/ui/webui/test/common/cdp.py", line 205, in invoke
    res = self.command(cmd)
  File "/root/workdir/anaconda/ui/webui/test/common/cdp.py", line 222, in command
    raise RuntimeError("CDP broken")
RuntimeError: CDP broken
Journal extracted to MyNewTest-test_default_installation-fedora-37-192.168.122.236-22-FAIL.log.gz
rsync: [sender] link_stat "/tmp/anaconda-tb-*" failed: No such file or directory (2)
rsync error: some files/attrs were not transferred (see previous errors) (code 23) at main.c(1823) [Receiver=3.2.3]
rsync: [Receiver] write error: Broken pipe (32)
Traceback (most recent call last):
  File "/root/workdir/anaconda/./ui/webui/test/integration/my_new_test.py", line 33, in test_default_installation
    self.run_integration_test()
  File "/root/workdir/anaconda/ui/webui/test/helpers/integration.py", line 109, in run_integration_test
    self._installer.open()
  File "/root/workdir/anaconda/ui/webui/test/helpers/step_logger.py", line 40, in wrapper
    result = function(*args, **kwargs)
  File "/root/workdir/anaconda/ui/webui/test/helpers/installer.py", line 92, in open
    self.browser.open(f"/cockpit/@localhost/anaconda-webui/index.html#/{step}")
  File "/root/workdir/anaconda/ui/webui/test/common/testlib.py", line 235, in open
    self._set_window_size(size[0], size[1])
  File "/root/workdir/anaconda/ui/webui/test/common/testlib.py", line 966, in _set_window_size
    self.cdp.invoke("Emulation.setDeviceMetricsOverride",
  File "/root/workdir/anaconda/ui/webui/test/common/cdp.py", line 205, in invoke
    res = self.command(cmd)
  File "/root/workdir/anaconda/ui/webui/test/common/cdp.py", line 215, in command
    self.start()
  File "/root/workdir/anaconda/ui/webui/test/common/cdp.py", line 327, in start
    self.invoke("Page.addScriptToEvaluateOnNewDocument", source=src, no_trace=True)
  File "/root/workdir/anaconda/ui/webui/test/common/cdp.py", line 205, in invoke
    res = self.command(cmd)
  File "/root/workdir/anaconda/ui/webui/test/common/cdp.py", line 222, in command
    raise RuntimeError("CDP broken")
RuntimeError: CDP broken

# Result test_default_installation (__main__.MyNewTest) failed
# 1 TEST FAILED [2s on 71256f31a8c3]


```